osquery/logger/plugins/kafka_producer.cpp

```diff
@@ -0,0 +1,236 @@
+  }
+}
+
+KafkaProducerPlugin::~KafkaProducerPlugin() {
+  stop();
+}
+
+void KafkaProducerPlugin::stop() {
+  if (running_) {
+    running_ = false;
+    // wait for max 3 seconds
+    rd_kafka_flush(producer_.get(), 3 * 1000);
+  }
+}
+
+void KafkaProducerPlugin::init(const std::string& name,
+                               const std::vector<StatusLogLine>& log) {
+  // Get local hostname to use as client id and Kafka msg key.
+  char hostname[sysconf(_SC_HOST_NAME_MAX)];
+  if (gethostname(hostname, sizeof(hostname)) != 0) {
+    LOG(ERROR) << "Could not get system local hostname.";
+    return;
+  }
+
+  msgKey_ = std::string(hostname) + "_" + name;
+
+  // Configure Kafka producer.
+  char errstr[512];
+  rd_kafka_conf_t* conf = rd_kafka_conf_new();
+
+  if (rd_kafka_conf_set(conf, "client.id", hostname, errstr, sizeof(errstr)) !=
+      RD_KAFKA_CONF_OK) {
+    LOG(ERROR) << "Could not initiate Kafka client.id configuration:" << errstr;
+    return;
+  }
+
+  if (rd_kafka_conf_set(conf,
+                        "bootstrap.servers",
+                        FLAGS_logger_kafka_brokers.c_str(),
+                        errstr,
+                        sizeof(errstr)) != RD_KAFKA_CONF_OK) {
+    LOG(ERROR) << "Could not initiate Kafka brokers configuration: " << errstr;
+    return;
+  }
+
+  // Register send callback.
+  rd_kafka_conf_set_dr_msg_cb(conf, onMsgDelivery);
+
+  // Create producer handle.
+  std::unique_ptr<rd_kafka_t, std::function<void(rd_kafka_t*)>> rk(
+      rd_kafka_new(RD_KAFKA_PRODUCER, conf, errstr, sizeof(errstr)),
+      delKafkaHandle);
+  producer_.swap(rk);
+  if (producer_.get() == nullptr) {
+    LOG(ERROR) << "Could not initiate Kafka producer handle: " << errstr;
+    return;
+  }
+
+  // Set topic configurations.
+  rd_kafka_topic_conf_t* topicConf = rd_kafka_topic_conf_new();
+  if (rd_kafka_topic_conf_set(topicConf,
+                              "acks",
+                              FLAGS_logger_kafka_acks.c_str(),
+                              errstr,
+                              sizeof(errstr)) != RD_KAFKA_CONF_OK) {
+    LOG(ERROR)
+        << "Could not initiate Kafka request.required.acks configuration: "
+        << errstr;
+    return;
+  }
+
+  // Initiate Kafka topic.
+  std::unique_ptr<rd_kafka_topic_t, std::function<void(rd_kafka_topic_t*)>> tk(
+      rd_kafka_topic_new(
+          producer_.get(), FLAGS_logger_kafka_topic.c_str(), topicConf),
+      delKafkaTopic);
+  topic_.swap(tk);
+  if (topic_.get() == nullptr) {
+    LOG(ERROR) << "Could not create Kafka topic " << FLAGS_logger_kafka_topic
+               << ": " << rd_kafka_last_error();
+    return;
+  }
+
+  // Start bg loop for polling to ensure onMsgDelivery callback is invoked even
+  // at times were no messages are produced
+  // (http://docs.confluent.io/2.0.0/clients/producer.html#asynchronous-writes)
+  running_ = true;
+  futureTimer_ = std::async(std::launch::async, [this]() {
```

Comment:
The scheduler is fairly basic: https://github.com/facebook/osquery/blob/master/osquery/dispatcher/scheduler.cpp#L148

Essentially:
```cpp
YourThing::start() {
  // do some work
  // Put the thread into an interruptible sleep,
  //   it will break out of the sleep when interrupted.
  pauseMilli(5000);
  if (interrupted()) {
    break;
  }
}
```